### PR TITLE
[v1alpha2] machines: Re-enable support for availability zones

### PR DESCRIPTION
**Pull request by justaugustus**
Signed-off-by: Stephen Augustus <saugustus@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: api/v1alpha2/azuremachine_types.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ const (
 
 // AzureMachineSpec defines the desired state of AzureMachine
 type AzureMachineSpec struct {
-	AvailabilityZone string `json:"availabilityZone"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
```
Comment (vincepri) [inline api/v1alpha2/azuremachine_types.go]: Is this required or optional?

File: controllers/azuremachine_reconciler.go [hunk 2]
```diff
@@ -215,8 +200,22 @@ func (s *azureMachineService) getVirtualMachineZone() (string, error) {
...
-	return zones[rand.Intn(len(zones))], nil
+	var selectedZone string
+	if zone != "" {
+		for _, allowedZone := range zones {
+			if allowedZone == zone {
+				selectedZone = zone
+			}
+		}
+	} else {
+		klog.Infof("selecting AZ at random: no availability zone set or selected availability zone is not supported for VM size %s in location %s", vmSize, location)
```
Comment (vincepri) [inline controllers/azuremachine_reconciler.go]:
```suggestion
		klog.Infof("Selecting AZ at random: no availability zone set or selected availability zone is not supported for VM size %s in location %s", vmSize, location)
```

File: controllers/azuremachine_reconciler.go [hunk 3]
```diff
@@ -215,8 +200,22 @@ func (s *azureMachineService) getVirtualMachineZone() (string, error) {
...
+		rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
+		selectedZone = zones[rand.Intn(len(zones))]
+	}
+
+	klog.Infof("selecting availability zone %s for %s", selectedZone, vmName)
```
Comment (vincepri) [inline controllers/azuremachine_reconciler.go]:
```suggestion
	klog.Infof("Selected availability zone %s for %s", selectedZone, vmName)
```

File: controllers/azuremachine_reconciler.go [hunk 4]
```diff
@@ -215,8 +200,22 @@ func (s *azureMachineService) getVirtualMachineZone() (string, error) {
...
+	if zone != "" {
+		for _, allowedZone := range zones {
+			if allowedZone == zone {
+				selectedZone = zone
+			}
+		}
+	} else {
+		klog.Infof("selecting AZ at random: no availability zone set or selected availability zone is not supported for VM size %s in location %s", vmSize, location)
+		rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
+		selectedZone = zones[rand.Intn(len(zones))]
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply controllers/azuremachine_reconciler.go]: Great point! Set this to `selectedZone = zones[0]` instead.

File: controllers/azuremachine_reconciler.go [hunk 5]
```diff
@@ -215,8 +200,22 @@ func (s *azureMachineService) getVirtualMachineZone() (string, error) {
...
+	var selectedZone string
+	if zone != "" {
+		for _, allowedZone := range zones {
+			if allowedZone == zone {
+				selectedZone = zone
+			}
+		}
+	} else {
+		klog.Infof("selecting AZ at random: no availability zone set or selected availability zone is not supported for VM size %s in location %s", vmSize, location)
+		rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply controllers/azuremachine_reconciler.go]: @devigned -- I removed the randomness altogether, based on @vincepri's [comment](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/278#discussion_r329785525).

File: api/v1alpha2/azuremachine_types.go (hunk 1 above)
Comment (justaugustus) [reply api/v1alpha2/azuremachine_types.go]: @vincepri -- I think leveraging AZs should be a goal as they provide a better guarantees than standard VMs.

The field isn't required, but the code prevents a foot-gun situation by choosing an AZ for you if you haven't already done so.

Comment (vincepri) [reply api/v1alpha2/azuremachine_types.go]: How about then to have something like this

```suggestion
        // <Add description here>
        // +optional
	AvailabilityZone *string `json:"availabilityZone,omitempty"`
```

This helps us having the nil case (not specified) as being the default one, without supplying the empty string

File: controllers/azuremachine_reconciler.go [hunk 6]
```diff
@@ -215,8 +199,21 @@ func (s *azureMachineService) getVirtualMachineZone() (string, error) {
 		return "", nil
 	}
 
-	rand.Seed(time.Now().Unix()) // initialize global pseudo random generator
-	return zones[rand.Intn(len(zones))], nil
+	var selectedZone string
+	if zone != "" {
+		for _, allowedZone := range zones {
+			if allowedZone == zone {
+				selectedZone = zone
```
Comment (vincepri) [inline controllers/azuremachine_reconciler.go]: Add a `break` after this?

**Comment by juan-lee** [comment]
Adding this post-merge to document how azure works when a zone is specified and not.

If a zone is specified in the vm api azure will colocate IPs, disks, and the vm to ensure zonal isolation. Azure will also configure associated load balancers to be zonally redundant. 

If a zone is not specified, IPs, disks, and vms may land in any availability zone. Load balancers won't be configured for zonal redundancy.

Specifying a single zone for all VMs in a cluster will result in the least zone redundant configuration. This means our defaults will result in a cluster that will be completely down if the zone it's in fails. We'll need to revisit this in a future PR.

Thanks for getting AZ re-enabled @justaugustus!!!

**Pull request merged**